### PR TITLE
[RU] Add Metaprogramming lesson

### PR DESCRIPTION
**Pull request by brain-geek**
This PR adds Russian translation of Metaprogramming lesson.

Hey, @nscyclone , can you please review this?

### CODE REVIEW

File: ru/lessons/advanced/metaprogramming.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+{:divide, [], [42, {:denominator, [], Elixir}]}
+iex> quote do: divide(42, unquote(denominator))
+{:divide, [], [42, 2]}
+```
+
+В первом примере переменная `denominator` экранирована, потому результирующий AST включает кортеж для доступа к переменной. Во втором примере результирующий код имеет в себе только значение этой переменной.
+
+## Макросы
+
+Как только мы разобрались как работают `quote/2` и `unquote/1`, мы готовы к погружению в макросы. Важно помнить, что макросы, как и все в метапрограммировании, должны использоваться очень осторожно.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: There should be a comma between `разобрались` and `как`.

File: ru/lessons/advanced/metaprogramming.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+В простейших терминах макросы - это специальные функции, созданные для возврата экранированных выражений, которые затем будут вставлены в код приложения. Представьте себе что макрос заменяется значением, которое он вернет, а не будет вызван как функция. С макросами у нас есть все необходимое для расширения языка и динамического добавления кода в наше приложение.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: Another comma is missing between `Представьте себе` and `что`.

File: ru/lessons/advanced/metaprogramming.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+А когда логирование будет отключено - вот такой:
+
+```elixir
+def test do
+end
+```
+
+### Закрытые макросы
+
+Хоть это и не частый случай, Elixir поддерживает закрытые макросы. Закрытый макрос определяется вызовом `defmacrop` и может быть вызван только из того модуля, в котором он был определен. Закрытые макросы должны быть определены раньше чем будут вызваны кодом.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: One more comma should be added after `определены раньше`.

File: ru/lessons/advanced/metaprogramming.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+### Гигиена при использовании макросов
+
+Как макросы взаимодействуют с контекстом вызова называется гигиеной макросов. По умолчанию, макрос в Elixir гигиеничен и не будет конфликтовать с контекстом:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: I'd rephrase the first sentence like `Взаимодействие макросов с контекстом вызова называется гигиеной макросов.`.
A comma in the second sentence is redundant.

File: ru/lessons/advanced/metaprogramming.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+nil
+iex> val = 42
+42
+iex> Example.hygienic
+-1
+iex> val
+42
+```
+
+Но что если мы хотим изменять значение `val`? Для того чтобы обозначить переменную негигиеничной мы можем использовать `var!/2`. Давайте изменим пример с использованием этого трюка:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: A comma should be between `что` and `если` in the first sentence and between `негигиеничной` and `мы` in the second one.

File: ru/lessons/advanced/metaprogramming.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    quote do: val = -1
+  end
+
+  defmacro unhygienic do
+    quote do: var!(val) = -1
+  end
+end
+```
+
+И посмотрим как он теперь взаимодействует с контекстом:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: There should be a comma after `посмотрим`.

File: ru/lessons/advanced/metaprogramming.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+-1
+iex> val
+42
+iex> Example.unhygienic
+-1
+iex> val
+-1
+```
+
+Включая `var!/2` в макрос мы изменили значение `val` не передавая его внутрь макроса. Использование таких макросов должно быть минимальным, ведь таким образом мы увеличиваем вероятность конфликта имен.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: Several commas are missing here. The first sentence should look like the following: 
>Включая `var!/2` в макрос, мы изменили значение `val`, не передавая его внутрь макроса.

File: ru/lessons/advanced/metaprogramming.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+### Контекст
+
+Мы уже рассмотрели полезность `unquote/1`, но есть еще один нюанс при включении значений в наш код: контекст. С контекстом переменных мы можем включить один код несколько раз внутри макроса и убедиться что они выполняться только один раз, препятствуя повторным выполнениям. Для использования этой возможности, названия таких переменных мы должны передать ключевым списком в опцию `bind_quoted` функции `quote/2`.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: `убедиться что они выполняться` probably should be `убедиться, что он выполнится` where `он` means `код`.
I'd rephrase the last sentence to make it sound more naturally:
>Для использования этой возможности мы должны передать названия таких переменных в опцию `bind_quoted` функции `quote/2` в виде ключевого списка.

File: ru/lessons/advanced/metaprogramming.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+  defmacro double_puts(expr) do
+    quote do
+      IO.puts unquote(expr)
+      IO.puts unquote(expr)
+    end
+  end
+end
+```
+
+Попробуем наш новый макрос передав ему функцию текущего системного времени. Мы ожидаем что она выведет значение дважды:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: Several commas are missing here: before `передав` and before `что она выведет`.

File: ru/lessons/advanced/metaprogramming.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+iex> require Example
+nil
+iex> Example.double_puts(:os.system_time)
+1450476083466500000
+1450476083466500000
+```
+
+С `bind_quoted` мы получаем ожидаемый результат: одно и то же время выведено дважды.
+
+Теперь, когда мы покрыли `quote/2`, `unquote/1`, и `defmacro/2`, у нас есть все необходимые инструменты для расширения языка Elixir под наши нужды.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: I suppose that `покрыли` is not very good in this case. What do you think about `разобрали` or `познакомились с`?

**Comment by nscyclone** [comment]
Thanks for the translation, @brain-geek! It looks fine but I've some issues there, mostly with punctuation.

**Comment by brain-geek** [comment]
Thanks for all this feedback, @nscyclone . It was tough lesson for me. :)

I've applied all your suggestions and changed few more parts. Can you review it once again?

File: ru/lessons/advanced/metaprogramming.md [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+Попробуем наш новый макрос,передав ему функцию текущего системного времени. Мы ожидаем, что она выведет значение дважды:
+
+```elixir
+iex> Example.double_puts(:os.system_time)
+1450475941851668000
+1450475941851733000
+```
+
+Время отличается! Что произошло? Использование `unquote/1` на одном и том же выражении несколько раз приводит к выполнению этого кода несколько раз, что может иметь неприятные последствия. Давайте обновим наш код с использованием `bind_quoted` и посмотрим что получается:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: How about changing `иметь неприятные последствия` to `обернуться неприятными последствиями`?
`посмотрим что получается` → `посмотрим, что получится`

File: ru/lessons/advanced/metaprogramming.md [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+  defmacro double_puts(expr) do
+    quote do
+      IO.puts unquote(expr)
+      IO.puts unquote(expr)
+    end
+  end
+end
+```
+
+Попробуем наш новый макрос,передав ему функцию текущего системного времени. Мы ожидаем, что она выведет значение дважды:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: A space is expected after the comma here.

File: ru/lessons/advanced/metaprogramming.md [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+-1
+iex> val
+-1
+```
+
+Включая var!/2 в макрос, мы изменили значение val, не передавая его внутрь макроса. Использование таких макросов должно быть минимальным, ведь таким образом мы увеличиваем вероятность конфликта имен.
+
+### Контекст
+
+Мы уже рассмотрели полезность `unquote/1`, но есть еще один нюанс при включении значений в наш код: контекст. С контекстом переменных мы можем включить один код несколько раз внутри макроса и убедиться, что он выполнится только один раз, препятствуя повторным выполнениям. Для использования этой возможности мы должны передать названия таких переменных в опцию bind_quoted функции quote/2 в виде ключевого списка.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: You probably forgot to insert the backticks around `bind_quoted` and `quote/2`.

File: ru/lessons/advanced/metaprogramming.md [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+А когда логирование будет отключено - вот такой:
+
+```elixir
+def test do
+end
+```
+
+### Закрытые макросы
+
+Хоть это и не частый случай, Elixir поддерживает закрытые макросы. Закрытый макрос определяется вызовом `defmacrop` и сможет быть вызван только из того модуля, в котором он был определен. Закрытые макросы должны быть определены раньше, чем будут вызваны кодом.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: `не частый` → `нечастый`

File: ru/lessons/advanced/metaprogramming.md [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+```elixir
+iex> require OurMacro
+nil
+iex> OurMacro.unless true, do: "Hi"
+nil
+iex> OurMacro.unless false, do: "Hi"
+"Hi"
+```
+
+Так как макрос заменяет код в самом приложении, мы можем контролировать что и когда компилируется. Отличным примером является модуль `Logger`. Когда логирование отключено, код не подключается и результирующее приложение не содержит никаких отсылок к коду логирования. Этим Elixir отличается от других языков, в которых будут накладные расходы на вызовы пустых функций.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: A comma is missing before `что` in the first sentence and before `и результирующее приложение` in the second.

File: ru/lessons/advanced/metaprogramming.md [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+```elixir
+iex> denominator = 2
+2
+iex> quote do: divide(42, denominator)
+{:divide, [], [42, {:denominator, [], Elixir}]}
+iex> quote do: divide(42, unquote(denominator))
+{:divide, [], [42, 2]}
+```
+
+В первом примере переменная `denominator` экранирована, потому результирующий AST включает кортеж для доступа к переменной. Во втором примере результирующий код имеет в себе только значение этой переменной.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: I'd suggest to replace `Во втором примере результирующий код имеет в себе только значение этой переменной.` with `Во втором примере в результирующем коде будет только её значение.`.

File: ru/lessons/advanced/metaprogramming.md [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+1
+iex> [1, 2] # Списки
+[1, 2]
+iex> {"hello", :world} # кортеж из 2 элементов
+{"hello", :world}
+```
+
+## Unquote
+
+Теперь, когда мы можем получить внутреннюю структуру кода, как ее изменить? Для вставки нового кода или значений мы используем `unquote/1`. Когда мы раскрываем выражение, оно будет выполнено и результат вставлен в AST. Для демонстрации этого давайте посмотрим на пару примеров:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: Not sure if the first sentence would sound better if rephrased like `Теперь мы можем получить внутреннюю структуру кода, но как ее изменить?`.

File: ru/lessons/advanced/metaprogramming.md [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+:world
+iex> quote do: 1 + 2
+{:+, [context: Elixir, import: Kernel], [1, 2]}
+iex> quote do: if value, do: "True", else: "False"
+{:if, [context: Elixir, import: Kernel],
+ [{:value, [], Elixir}, [do: "True", else: "False"]]}
+iex(6)>
+```
+
+Заметили что первые три не возвращают кортеж? Есть всего пять выражений, которые возвращают сами себя при использовании `quote/2`:
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: One more comma is missing before `что`.

File: ru/lessons/advanced/metaprogramming.md [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+
+Метапрограммирование - процесс написания кода с помощью другого кода. В Elixir это дает нам возможность расширять возможности языка для большего соответствия нуждам приложения. Мы начнем этот урок с рассмотрения как код Elixir представлен внутри, а потом - как его модифицировать. И, в итоге, мы применим эти знания для создания собственных расширений языка.
+
+Предупреждение: метапрограммирование это сложно и оно должно использоваться только когда абсолютно необходимо. Злоупотребление этим инструментом приведет к нечитабельности и сложностям в отладке кода.
+
+{% include toc.html %}
+
+## Quote
+
+Первый шаг к пониманию метапрограммирования - понимание что представляют собой выражения. В Elixir внутреннее представление абстрактного синтаксического дерева (AST) состоит из кортежей. Эти кортежи состоят из трех частей: название функции, метаданные и аргументы функции.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: Another comma is missing before `что`.

File: ru/lessons/advanced/metaprogramming.md [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+layout: page
+title: Метапрограммирование
+category: advanced
+order: 7
+lang: ru
+---
+
+Метапрограммирование - процесс написания кода с помощью другого кода. В Elixir это дает нам возможность расширять возможности языка для большего соответствия нуждам приложения. Мы начнем этот урок с рассмотрения как код Elixir представлен внутри, а потом - как его модифицировать. И, в итоге, мы применим эти знания для создания собственных расширений языка.
+
+Предупреждение: метапрограммирование это сложно и оно должно использоваться только когда абсолютно необходимо. Злоупотребление этим инструментом приведет к нечитабельности и сложностям в отладке кода.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: A comma is missing before `когда`.

File: ru/lessons/advanced/metaprogramming.md [hunk 21]
```diff
@@ -0,0 +1,238 @@
+---
+layout: page
+title: Метапрограммирование
+category: advanced
+order: 7
+lang: ru
+---
+
+Метапрограммирование - процесс написания кода с помощью другого кода. В Elixir это дает нам возможность расширять возможности языка для большего соответствия нуждам приложения. Мы начнем этот урок с рассмотрения как код Elixir представлен внутри, а потом - как его модифицировать. И, в итоге, мы применим эти знания для создания собственных расширений языка.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: A comma should be added before `как код`.

**Comment by nscyclone** [comment]
@brain-geek, I've just looked through it once more and found some more shortcomings.

**Comment by brain-geek** [comment]
@nscyclone Your review is very useful! Thanks, I've applied all the changes.

File: ru/lessons/advanced/metaprogramming.md [hunk 22]
```diff
@@ -0,0 +1,238 @@
+---
+layout: page
+title: Метапрограммирование
+category: advanced
+order: 7
+lang: ru
+---
+
+Метапрограммирование - процесс написания кода с помощью другого кода. В Elixir это дает нам возможность расширять возможности языка для большего соответствия нуждам приложения. Мы начнем этот урок с рассмотрения внутреннего представления Elixir кода, а потом - как его модифицировать. И, в итоге, мы применим эти знания для создания собственных расширений языка.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: Maybe we could improve the third sentence here: I think a verb would fit fine before: `как его модифицировать`. What do you think about `а потом узнаем, как его можно модифицировать.`?

**Comment by nscyclone** [comment]
@brain-geek, I've found one more possible improvement but it's up to you, everything else is fine. Thanks a lot for your contribution!

File: ru/lessons/advanced/metaprogramming.md (hunk 22 above)
Comment (brain-geek) [inline ru/lessons/advanced/metaprogramming.md]: I've applied this, but a bit different.

**Comment by brain-geek** [comment]
@nscyclone That part is now changed too. What do you say now about the PR? :)

File: ru/lessons/advanced/metaprogramming.md [hunk 23]
```diff
@@ -0,0 +1,238 @@
+---
+layout: page
+title: Метапрограммирование
+category: advanced
+order: 7
+lang: ru
+---
+
+Метапрограммирование - процесс написания кода с помощью другого кода. В Elixir это дает нам возможность расширять возможности языка для большего соответствия нуждам приложения. Мы начнем этот урок с рассмотрения внутреннего представления Elixir кода. А уже потом - узнаем как его модифицировать. И, в итоге, мы применим эти знания для создания собственных расширений языка.
```
Comment (nscyclone) [inline ru/lessons/advanced/metaprogramming.md]: I suppose the hyphen is not needed in the fourth sentence but a comma is missing before `как`. Also it probably would be better to merge this sentence with the previous one using a comma between them. So finally it could look like `Мы начнем этот урок с рассмотрения внутреннего представления Elixir кода, а потом узнаем, как его модифицировать`.

**Comment by nscyclone** [comment]
@brain-geek, it looks good but I've left a note about the sentence you fixed in the latest commit. I think it will be ready to merge after finishing with this one.

**Comment by brain-geek** [comment]
@nscyclone I've changed to your version. Thanks for patience!

**Comment by nscyclone** [comment]
@brain-geek, I am not sure whether you expect me to review this once more but it seems perfect now and is ready to be merged. Thanks a lot!

**Comment by brain-geek** [comment]
@nscyclone Thanks for the signoff. Yes, I was waiting for it. :)
